Improve batched torch.matrix_exp performance

scruel opened the pull request:
Fixes #105225

- New implementation for `compute_T18_scale_square` method.
- Always use the new highest degree algo for large batch sizes (size > 1).


cc @jianyuh @nikitaved @pearu @mruberry @walterddr @xwang233 @Lezcano @JubilantJerry

scruel committed:
Fix: `torch.matrix_exp` performance issue (#105225)

scruel committed:
Fix: prevent to mask norm on CPU for large batch size

lezcano commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -2546,13 +2569,22 @@ Tensor mexp_impl(
-  // based on their norm. This decision takes place on CPU.
-  // It requires moving data back and forth between devices when `a` is on CUDA,
-  // but at the cost of only one sigle CPU-CUDA synchronization (instead of 6),
+  const auto norm_max = norm.max().to(at::kCPU).item().toFloat();
```
```suggestion
  const auto norm_max = norm.max().to(at::kCPU).item<float>();
```

scruel replied:
Done.

lezcano commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -2165,6 +2172,7 @@ using array2d = std::array<std::array<scalar_t, COL>, ROW>;
 // we consider 6 Taylor expansions of degree
 // 1, 2, 4, 8, 12, 18
 constexpr int total_n_degs = 6;
+constexpr int large_batch_threshold = 100;
```
Benchmarks backing up this number?

lezcano commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -2546,13 +2569,22 @@ Tensor mexp_impl(
+  // expensive (compared to the synchronization overhead from GPU to CPU), so
+  // here we have a threshold to determine whether or not to move norm to CPU.
+  const auto norm_small_to_cpu = ((a.device().type() == at::kCUDA) 
+    && a.size(0) >= large_batch_threshold) ? norm.to(at::kCPU) : norm;
```
Here you are sending large batches to CPU, which is the opposite to what you claim you are doing. Please provide benchmarks that back up this optimisation.

lezcano commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -2517,25 +2525,40 @@ void compute_T18_scale_square(
-  ).unsqueeze(-1).unsqueeze(-1).to(at::kLong);
-  const auto pow2s = at::pow(2, s);
-  const auto a_scaled = a / pow2s;
+  const auto s = (at::floor(at::log2(norm / theta)) + 1).clamp(/*min=*/0);
```
Mind adding a small description of the algorithm on top (the sorting and all that stuff)

lezcano reviewed (commented):
The implementation looks good. I would like to see some benchmarks for a number of decisions though. When benchmarking, we would want to find a set of matrices that is representative. Matrices where all the matrices in the batch have the same norm for this algorithm, one where the norms are varied, etc.

On a related point, setting `compute_highest_degree_approx=True` on CUDA (perhaps also on CPU) will likely yield reasonable speed-ups, as it'll avoid plenty of nonzero calls. We should benchmark this

scruel committed:
Fix wrong slice length issue & add code comments

lezcano commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -2517,25 +2525,65 @@ void compute_T18_scale_square(
+  // get `s` as [4, 1, 1, 0], so we can use it to get the result by calculating
+  // matrix[0]^(2^4), matrix[1]^(2^1) and matrix[2]^(2^1) one by one to get the
+  // result, such "one by one calculation" will be quite slow.
+  const auto s = (at::floor(at::log2(norm / theta)) + 1).clamp(/*min=*/0);
```
Why changing ceil to floor + 1? Also, we should probably cast s to an integer, as it was done before.

lezcano commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -2517,25 +2525,65 @@ void compute_T18_scale_square(
-  const auto pow2s = at::pow(2, s);
-  const auto a_scaled = a / pow2s;
+  // We eventually need to do the matrix multiplication to calculate the result.
+  // For exampe, if we have `norm` equal to [27, 6, 6, 0.05], we will end up to
```
nit. example

The comments have quite a few errata. Mind fixing those?

scruel replied:
Oh, sorry, but can you be more clarify?

lezcano commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -2517,25 +2525,65 @@ void compute_T18_scale_square(
+  // Compose the result back
+  auto output = at::cat(output_pieces, 0);
+  output = output.index_select(0, at::argsort(sorted_s_inds));
+  mexp_out.copy_(output);
```
Passing `mexp_out` does not make sense anymore, as we can simply return `index_select` here rather than performing an unnecessary copy.

lezcano commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -2517,25 +2525,65 @@ void compute_T18_scale_square(
+  auto mexp_scaled = at::native::compute_T18<scalar_t>(a_scaled);
+
+  // Sort:
+  // Consider inputs are squre matrix, so if we first power `martix 0,1,2`, then
```
square

lezcano commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -2517,25 +2525,65 @@ void compute_T18_scale_square(
+
+  // Sort:
+  // Consider inputs are squre matrix, so if we first power `martix 0,1,2`, then
+  // the ramain thing will only be multiply `matrix 0` by (2^4 - 1) times, which
```
remain

lezcano replied:
there's plenty of "matrixs" as well. Running the whole thing through a grammar check would certainly help

scruel replied:
Sure, will fix those typos, thanks

scruel committed:
Remove mexp_out to return directly & clean code

scruel committed:
Set `large_batch_threshold` based on the brenchmark result

scruel commented:
[------------- matrix_exp on large batch -------------]
                           |    cpu      |   cuda:0
64 threads: ------------------------------------------
      shape (100, 8, 8)    |  4.5/5      |  4.8/4.9
      shape (200, 8, 8)    |  4.6/4.61   |  4.75/4.7
      shape (300, 8, 8)    |  4.63/4.8   |  4.7/4.8
      shape (400, 8, 8)    |  4.72/4.79   |  4.8/4.6
      shape (500, 8, 8)    |  4.91/4.96  |  4.7/4.8
      shape (1000, 8, 8)   |  5.6/5.6    |  4.7/4.6
      shape (10000, 8, 8)  |  17.3/17.4  |  4.61/4.6

Times are in milliseconds (ms).

Benchmarked env info:
NVIDIA A100-SXM4-40GB
AMD EPYC 7543 32-Core Processor
CUDA used to build PyTorch: 11.6
Nvidia driver version: 525.125.06

scruel commented:
Benchmark result:
(Debug build, Result value meaning: Random norms/Same norms)

1. With `compute_highest_degree_approx=false`:
```
[-------------------- matrix_exp --------------------]
                           |    cpu       |   cuda:0
64 threads: ------------------------------------------
      shape (1, 2, 2)      |   4.1/4      |   4.5/4.7
      shape (100, 2, 2)    |   6.8/4.5    |   7.3/4.9
      shape (200, 2, 2)    |   5.5/5      |   6/4.9
      shape (300, 2, 2)    |   7/4.7      |   7.1/4.8
      shape (400, 2, 2)    |   7/4.9      |   7/4.9
      shape (500, 2, 2)    |   7.5/5      |   7.1/4.9
      shape (1000, 2, 2)   |   8/6        |   7/4.9
      shape (10000, 2, 2)  |   20/17      |   7/5

Times are in milliseconds (ms).
```

2. With `compute_highest_degree_approx=true`:
```
[-------------------- matrix_exp --------------------]
                             |    cpu      |   cuda:0
64 threads: ------------------------------------------
      shape (1, 2, 2)        |   3/3.3     |   3/3.2
      shape (100, 2, 2)      |   3/3.2     |   3/3.2
      shape (200, 2, 2)      |   2.9/3.2   |   3/3.2
      shape (300, 2, 2)      |   2.9/3.2   |   3/3
      shape (400, 2, 2)      |   3/3       |   2.9/3.2
      shape (500, 2, 2)      |   3.2/3.3   |   3/3.2
      shape (1000, 2, 2)     |   3.1/3     |   2.9/3.2
      shape (10000, 2, 2)    |   2.9/3.3   |  2.8/3.1
      shape (100000, 2, 2)   |   3.3/3.4   |  3.1/3.2
      shape (1000000, 2, 2)  |   12.2/18.3 |  12/18

Times are in milliseconds (ms).
```

====
Env info:
NVIDIA A100-SXM4-40GB
AMD EPYC 7543 32-Core Processor
CUDA used to build PyTorch: 11.6
Nvidia driver version: 525.125.06


Code:
```
import torch
from torch.utils.benchmark import Timer, Compare

sizes = [1, 100, 200, 300, 400, 500, 1000, 10000, 100000, 1000000]

torch.manual_seed(1)

def gen_inputs():
    for xt in (x1, x2):
        yield torch.matrix_exp, xt


def benchmark(f, xt):
    torch.testing.assert_close(f(xt[1][:8]), torch.matrix_exp(xt[1][:8]))
    return Timer("f(xt[1])",
                 globals=locals(),
                 label="Norm",
                 sub_label=f"shape {tuple(xt[1].shape)}",
                 description=xt[0],
                 num_threads=torch.get_num_threads()).blocked_autorange(min_run_time=2)


for s in sizes:
    x1 = ('Random', torch.randn(s, 2, 2, device="cuda"))
    x2 = ('Same', torch.tensor([[1, 2.],[3., 25.]], device="cuda").repeat((s, 1, 1)))

    results = []
    for args in gen_inputs():
        results.append(benchmark(*args))

    compare = Compare(results)
    compare.trim_significant_figures()
    compare.print()
```

lezcano commented:
Can you also run similar experiments for large matrices, as the most normal use case for this function will be on larger matrices.
Also, in general, it's best to use batch sizes and sizes in general that are powers of two, as that will be the most common use case in practice when looking for raw speed.

scruel commented:
Cleaned benchmark result: 
https://pastebin.ubuntu.com/p/F85x4H4krF/

Benchmark testing code: https://pastebin.ubuntu.com/p/5vcttQ5mVM/
Note: because I failed to modify the function signature for `matrix_exp` to compile, so in the code, I just hacked a little bit with input martix with the following workaround in C++:

```cpp
mexp_impl...
...
  compute_highest_degree_approx = a[0][0][0].item<bool>();
  int use_cpu = a[0][0][1].item<bool>();
  const auto norm_small_to_cpu = (use_cpu == 1) ? norm.to(at::kCPU) : norm;
```

Full benchmark log:
- New Implementation:
https://pastebin.ubuntu.com/p/htQgQddsdr/
https://pastebin.ubuntu.com/p/BdxZbwdskd/
- Old Implementation:
https://pastebin.ubuntu.com/p/bTjWzDtTK5/
https://pastebin.ubuntu.com/p/WnMS2PQZS6/


Consider the problems I met while benchmarking, I wonder if you already have some wiki pages which I can read to help me add functions or modify function's signature?
And why you keep `compute_highest_degree_approx` as an inner parameter, I can't find anywhere to change its value via python code.
Furthermore, how about we can set this value automatically for users by `batch_size` and `torch.quantile(norm, ...)`?
BTW, I think we need to move `compute_highest_degree_approx` related discussion to a new issue.

scruel committed:
Adjust threshold & move unique_s to CPU

lezcano commented:
To benchmark the best way is to compile a branch, benchmark it, save the results, compile the other and repeat.

lezcano commented:
You can also merge the old and new results via Compare after loading them from a file, of course

lezcano commented:
If done properly, Compare will create the full table for you and you shouldn't need to do any of the manual editing that I guess you did for these benchmarks

scruel commented:
> To benchmark the best way is to compile a branch, benchmark it, save the results, compile the other and repeat.

I am using the dev branch (bb0fb95) and main branch (old, 5cc1a38) for benchmarking, for clarifying.
For small batch sizes, the new impl can be slower than the old impl, because the NaN check we added.

scruel commented:
> Generally, the benchmarks are a bit difficult to read. It's not clear whether the inputs are on CPU or CUDA. I'd assume you are just testing CUDA inputs here as you discuss moving the mask to CPU.

Oh, sorry, I already noticed this, so already add some notes about it, and it seems that it's still not quite enough, what we are testing is about the value of `large_batch_threshold`, which determine whether we should move `norms` to CPU before masking it.

> is not great either, as you are incurring in two synchronizations so the results are not going to be like those from the actual PR.

Yeah, but it's the only workaround I found currently (tried to modify the `matrix_exp` signature, but failed to compile), to avoid compiling by multiple times, and the side effect should be quite slight.

> If done properly, the scripts should output the table formatted for you and you should not need to do any postprocessing.
We indeed have two scripts, one for generating the result (full log), one for parsing the result into the cleaned version, the benchmark script was modified based on your comment (and the official document).

> it was an artifact of the initial implementation. I would like to remove it and just keep the compute_highest_degree_approx==True branch if we can prove that is indeed faster.
> At any rate, thank you for all this effort. I reckon that we are going to be able to speed-up this implementation quite a bit with this new algorithm :)
Oh, I see, I'm glad we can help~

lezcano commented:
Recompiling is the only reliable way of testing both of the configs in this case. Adding spurious synchronizations is not great. 

Could you please resubmit the benchmarks taking into account the points above? To reduce the running time of the benchmarks, you can remove the `min_run_time=2` from Timer, and reduce the number of configurations we test, as we are currently testing quite a few.

scruel commented:
> Recompiling is the only reliable way of testing both of the configs in this case. Adding spurious synchronizations is not great.
> 
> Could you please resubmit the benchmarks taking into account the points above? To reduce the running time of the benchmarks, you can remove the `min_run_time=2` from Timer, and reduce the number of configurations we test, as we are currently testing quite a few.

Sure, I can do that, but it will take some time, however, can you give me the `shape` range for reducing the number of configurations? thanks.

scruel commented:
> I would like to remove it and just keep the compute_highest_degree_approx==True branch if we can prove that is indeed faster.

However, we believe that there must be some cases where using the highest degree is faster, and there are cases where using it is slower.

lezcano commented:
What are examples of cases where using it is slower? It looks like it's generally faster in your benchmarks, isn't it?

scruel commented:
> What are examples of cases where using it is slower? It looks like it's generally faster in your benchmarks, isn't it?

How about:
```ipython
# Tensor on CPU
torch.manual_seed(1);v = torch.randn(1, 4096, 4096) * 0.0001;torch.matrix_exp(v);
%timeit torch.matrix_exp(v);

# Tensor on GPU
torch.manual_seed(1);v = torch.randn(1, 4096, 4096, device='cuda') * 0.0001;torch.matrix_exp(v);
%timeit torch.matrix_exp(v);

# Small tensor
torch.manual_seed(1);v = torch.randn(1, 4, 4, device='cuda') * 0.0001;torch.matrix_exp(v);
%timeit torch.matrix_exp(v);
```

Result:
https://pastebin.ubuntu.com/p/MT9vjNKG7z/


> For example, you could have 4 tables for {CPU, CUDA} x {same, ramdom}. Each table with rows with the size, and with columns "old, {mask_cpu, mask_cuda} x {highest_degree=True, highest_degree=False}" for example.

Benchmark Result:
https://pastebin.ubuntu.com/p/D2kkg2sHCZ/
Benchmark Code:
https://pastebin.ubuntu.com/p/G7w9GPgVFW/

scruel commented:
> there are some weird results on the input CPU table, on the LCPU column. For example for (65536, 2, 2) we get a time of 358414.7, but when we double the shape, the algorithm seems to run 10x faster??
> 
> What's going on there

It's really weird, this might be caused by the side effects of another colleague's program that was using quite a log CPU recourses at the same time, however, I tried to rerun the script and set batch sizes to $2^16$ ~ $2^19$, I couldn't reproduce such result.
I think I will need to rerun the benchmark after his script finished, sorry for the confusing.

scruel commented:
> I would like to remove it and just keep the compute_highest_degree_approx==True branch if we can prove that is indeed faster.
> Also, for batchsize = 1, we should probably always dispatch it to the previous algorithm, as there is no point running the new one on it.

BTW, before you consider to "remove" the old algorithm, I think that you could be interested about the commit we already pushed in another temp branch which "can" do the auto switch between the old one and the new one:
https://github.com/scruel/pytorch/commit/e273d404191e92d9560f2bfea309be18b84967c4

lezcano commented:
I wouldn't want to add another hyperparamenter, as hyperparameters work in some cases but not in others.

I get the feeling that, as it stands, you can do:
- If there's one matrix, perform the algorithm from the paper
- For more than one matrix, always compute the highest degree approximation

In most cases, the highest degree approximation will do the scale squaring. This will give us a pretty good algorithm overall.

scruel commented:
> I wouldn't want to add another hyperparamenter, as hyperparameters work in some cases but not in others.

Agree, that's why I didn't push that commit into this PR, however, which we are seeking for, is the "best" solution.

scruel commented:
> What's going on there

I think I may explain why, consider that new random matrix generate code we wrote:
```python
...
  t = torch.randn(batch_size, matrix_size , matrix_size, device=ADevice)
  t = torch.tril(t) - torch.tril(t).transpose(1, 2)
  return t * torch.exp((15 * torch.rand(batch_size, 1, 1, device=ADevice) - 5)) / matrix_size 
```

If we double the batch size, the norms can end to be quite smaller, so if we won't compute the highest degree approximation, we could get the faster results. But 10x? Still wired.

![img](https://github.com/pytorch/pytorch/assets/16933298/f4a63c29-3fab-44fa-a4cd-0b69391a0ac1)

Though it could also be a merge bug, even after a few checks, I couldn't find any bug.
For generating the previous result, I didn't expect that the threshold to have such a significant difference from the DEBUG build to the RELEASE build, so at the first, in order to reduce the number of test cases, I only set the range of batch sizes to 0 ~ 2^16. After I got the results, I decided to append the range 2^17 ~ 2^18, then merge the results into one as the final benchmark results (I confirmed that they are correct by checking the data separately).

No matter how, I will rerun the benchmark (avoiding the mentioned issues), need to have some patient to wait a little bit more time for the benchmark results :)

scruel commented:
> What's going on there

I think I may explain why, consider that new random matrix generate code we wrote:
```python
...
  t = torch.randn(batch_size, matrix_size , matrix_size, device=ADevice)
  t = torch.tril(t) - torch.tril(t).transpose(1, 2)
  return t * torch.exp((15 * torch.rand(batch_size, 1, 1, device=ADevice) - 5)) / matrix_size 
```

If we double the batch size, we can end up to ge a matrix which has quite small norms, so if we won't compute the highest degree approximation, we could be able to get the faster results. But 10x? Still wired.
And for same norms matrix, the results may have been affected by other programs:
![2023-10-13_03-39-36](https://github.com/pytorch/pytorch/assets/16933298/18e6b6be-e83e-424b-bffe-7e8a99e3a29a)

Although it could also be a merge bug, even after a few checks, I couldn't find any bug.
For generating the previous result, I didn't expect that the threshold to have such a significant difference from the DEBUG build to the RELEASE build, so at the first, in order to reduce the number of test cases, I only set the range of batch sizes to 0 ~ 2^16. After I got the results, I decided to append the range 2^17 ~ 2^18, then merge the results into one as the final benchmark results (I confirmed that the data are correct by checking them separately).

No matter how, I will check more and rerun the benchmark (avoiding the mentioned issues above), need to have some patient to wait a little bit more time for the benchmark results :)

JubilantJerry commented:
There may be some artifacts from using a debug build for benchmarking. Before getting more official benchmark numbers on an optimized build, I can try to sort out the theory for how the algorithm should be dispatched depending on the batch size, the matrix size, and the norm.

* matrix_exp should have runtime complexity of O(batch_size * matrix_size ** 3).
* Of this, the calls to Tn dominate the theoretical complexity. It should be that T18 is more expensive than the smaller degree functions, though the cost differs by at most a constant factor.
* The surrounding logic regarding norms, masks, unique consecutive, etc. at most have runtime complexity O(batch_size). The sorting and argsorting of the norms has runtime complexity O(batch_size * log(batch_size)).
* Of course, there are big constants involved and synchronization is especially expensive compared to other operations.

I've made the following logical deductions:

* If matrices are big enough, it's important to use the lowest degree approximation possible, since the Tn calls will dominate the cost, and of them T18 is the most expensive. This should be true regardless of batch size; using an unnecessarily large approximation for even one matrix in a batch would be a significant loss.
* If matrices are small, the Tn calls do not dominate the cost in theory. The sorting and argsorting in T18_scale_square would have the highest asymptotic cost. However, it takes a batch size in the millions for the cost of sorting to catch up with even one batch-matmul call for 2x2 matrices. In practice, for large batch sizes, batch-matmul is still the most expensive operation, assuming all operations are done on the GPU.
    * It is actually possible to do the sorting in linear time using fancy indexing, because we are rounding the log2 of the norms to integers and there is a reasonably small upper bound, but I think doing that would be an over-optimization so let's not.
* The cost of the batch-matmul calls increases with the degree of the approximation. So if the batch size is large enough, we again should still use the lowest degree approximation possible. The difference between the lower and higher degree approximations are more modest, though. If the norms of the matrices are such that we already were going to use T18 most of the time, then it would be wasteful to do masking and select the approximation degree based on the norms. We can skip directly to the highest degree approximation.
* For small matrix size and small batch size, the most significant cost is probably the synchronizations. I think there must be at least one synchronization, whether it's for choosing the approximation degree or if it's for counting the number of squaring steps. We should just skip directly to the highest degree approximation, which guarantees exactly 1 synchronization for all cases.

Therefore, I think the algorithm should be the following:
* Calculate the norms.
* If both batch size and matrix size is small, use an alternative method for NaN guarding that doesn't need synchronization. Then skip directly to the highest degree approximation. 
* Get the median of the norms, and synchronize it to CPU. Use it for NaN guarding.
* If the matrix size is small, and the median is above the threshold for using T18, skip directly to the highest degree approximation.
* Otherwise, choose the degree based on the norms and dispatch to the lowest acceptable degree approximation. We keep the norms on the GPU, and therefore use 7 synchronizations.

scruel commented:
Now the benchmark results about the tensors on CPU should be reasonable, I think the previous results were polluted by other programs that occupied CPU.

Benchmark with RELEASE build, on dev branch ([bb0fb95](https://github.com/pytorch/pytorch/commit/bb0fb956fc205907fe629347131fedea8a733ef9)) and main branch (old, [5cc1a38](https://github.com/pytorch/pytorch/commit/5cc1a38370f17b5d0353426a505fb15d7ab58322))

Benchmark Result:
https://pastebin.ubuntu.com/p/Xy4WjfyStN/

Benchmark Code (Unchanged):
https://pastebin.ubuntu.com/p/YxRw36zNTB/

scruel commented:
Now the benchmark results about the tensors on CPU should be reasonable, I think the previous results were polluted by other programs that occupied CPU.

Benchmark with RELEASE build, on dev branch ([bb0fb95](https://github.com/pytorch/pytorch/commit/bb0fb956fc205907fe629347131fedea8a733ef9)) and main branch (old, [5cc1a38](https://github.com/pytorch/pytorch/commit/5cc1a38370f17b5d0353426a505fb15d7ab58322))

Benchmark Result:
https://pastebin.ubuntu.com/p/Xy4WjfyStN/

Benchmark Code (Unchanged):
https://pastebin.ubuntu.com/p/YxRw36zNTB/

lezcano commented:
Big-o notation does not have much meaning on these algorithms. The reason is that big-o notation does not take into account parallelisation, and these algorithms are as fast as you can parallelise them. For context, SVD and matmul are both O(n³). Even worse, it does not take into account constants. This means, in particular, that it does not take into account how many times you are reading/writing from memory. Reading from memory takes about 200+ times longer than an arithmetic operation, so the big-o notation may very well be off by a factor of 200x.

All this is just talking about CPU. When dealing with heterogeneous computing, you also have to take into account synchronisations and copies from CPU to CUDA memory and back, which are particularly expensive.

In general, it's not good to overthink about heuristics, unless you can mathematically prove them to be correct. Otherwise, you will end up with an algorithm that very unpredictable performance characteristics and, as you are doing extra computations  in the basic case, it won't even be fast in the simpler cases. An example of this in your reasoning @JubilantJerry is when you say that we could use an insertion sorting to do the sorting. Sure, this can be done in linear time, but I bet you will not be able to implement it as fast as cub's sort without writing a specialised CUDA kernel and doing plenty of profiling. Not even close.

As per the benchmarks, I think we should implement what was discussed in https://github.com/pytorch/pytorch/pull/110848#issuecomment-1760152473. That would give us a pretty good performance in the 1 matrix case (very common) and it would always be faster than the other option. We would just lose a 5-10% speed-up on extremely large batches, which I expect to be very rare, so that's completely fine.

JubilantJerry commented:
Agreed that Big O is not necessarily the best thing to count on here. But even in the 1 matrix case it could be faster to use the highest degree approximation, and with batch size > 1 it could be faster to choose a lower degree approximation.

With the scheme proposed in #110848, there is also a sudden change in performance between 1 matrix and 2 matrices, and it creates a regression vs. the older algorithm. For example, consider a matrix of shape (2, 4096, 4096) and low norm. The older algorithm uses 2 low degree approximation calls, while the proposed method uses T18.

I realize my earlier pseudocode is too complicated though, I just about to simplify it. I edited the comment right before seeing the reply. The method is the following:

* Calculate est_cost = batch_size * matrix_size ** 2
* If est_cost < cost_threshold, skip directly to the highest degree approximation.
* Otherwise, choose the degree based on the norms and dispatch to the lowest acceptable degree approximation. We keep the norms on the GPU, and therefore use 7 synchronizations.

My simplified description has 1 hyperparameter (cost_thres) and I think it uses a better justified mechanism for choosing the algorithm to mostly ensure no regressions against the older algorithm.

lezcano commented:
Pretty much every function in `torch.linalg` has a drop in performance when you go from 1 to 2 matrices. That is completely fine as not many people use batches of such a small size.

JubilantJerry commented:
@scruel Thanks for the benchmark results. I think it can be improved with the following ideas:
1. Perform `torch.random.manual_seed` inside `get_matrix`. Create the matrices in the CPU, then convert the final result to ADevice. I think for random norms, the table has some artifacts where the different versions used different matrices, whose norms are very different. That's why there's a lot of variation for, e.g. shape (2, 4096, 4096) and shape (4, 4096, 4096)
2. I think we can try a Small Same Norm case. The current Same Norm case is norm 25, so it would use T18.

scruel committed:
Reduce one synchronization for the `compute_T18_scale_square` function.

scruel committed:
Remove the NaN check.

NaN check is mainly for preventing undefined behaviors while the inputs have NaN values, however, it can affect the performance for small batches.
In previous versions, the inputs' `nan` value will be gone after we perform `mexp`, which causes the NN to ouput the undefined "good" result rather than the obviously wrong result that includes NaN, or raise errors.

lezcano commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -2545,14 +2592,20 @@ Tensor mexp_impl(
   bool compute_highest_degree_approx = false
 ) {
   auto res = at::empty_like(a);
+  res.index_put_({at::indexing::Slice(), 0, 0}, std::numeric_limits<double>::quiet_NaN());
```
This doesn't make sense.

scruel replied:
NaN check is mainly for preventing undefined behaviors while the inputs have NaN values, however, it can affect the performance for small batches.
In previous versions, the inputs' `nan` value will be gone after we perform `mexp`, which causes the NN to output the undefined "good" result rather than the obviously wrong result that includes NaN, or raise errors.

This line is for keeping the NaN thing.

lezcano replied:
let's roll back all the nan-related fixes and keep this PR just about performance.
Then, you can open a second PR with a fix and a regression test fixing any NaN issue.

scruel replied:
Ok, it should be.

lezcano reviewed (commented):
Ok, I think that the discussion is deviating too much from the initial proposal. Alas, I don't have enough bandwidth as to continue shepherding this PR for much longer, so let's try to find a constructive way forward.

As it stands, I see two possibilities:
1. We merge it as-is after reverting the last `index_put_` edit.
2. We go with https://github.com/pytorch/pytorch/pull/110848#issuecomment-1760152473

Either is better than the current state, so I'm fine with either.

scruel commented:
> 1. We merge it as-is after reverting the last index_put_ edit.

How about the value of `large_batch_threshold`? We haven't yet decided after got the benchmarks.
It's for replacing the NaN check, I explained more in response to the above review.

> Either is better than the current state, so I'm fine with either.

@JubilantJerry What about you?

lezcano commented:
Pick anything you think it's reasonable. The speed-up you get in any direction is about 10% at best, and in most practical cases (n=256,512, b=128) is pretty much equivalent, so it doesn't matter all that much.

Of course, option 2 gets rid of this hyperparameter altogether, but yeah.

scruel committed:
Fix copy/paste issue :)

scruel committed:
Revert unnecessary changes & prepare for merging.

lezcano commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -2546,15 +2592,21 @@ Tensor mexp_impl(
 ) {
   auto res = at::empty_like(a);
```
Now this is not necessary for the `highest_degree_approx=True`

scruel committed:
Var res is not necessary for the highest_degree_approx=True condition

scruel committed:
fix typo :)

lezcano reviewed (approved):
LGTM

lezcano commented:
And thank you for all the benchmark and the thorough follow up :)

lezcano commented:
@pytorchbot merge

scruel commented:
> And thank you for all the benchmark and the thorough follow up :)

Thank you as well, you are really kind for answering all above questions, even some of them are quite tiny. :)

lezcano commented:
The CI failures are real tho.

scruel committed:
Fix indent

scruel commented:
> The CI failures are real tho.

The CI report is unrelated to this PR, I just checked the code again and fixed a wrong indent, even `make lint` doesn't count it as an error.
And I wonder is there anything else I can do?

lezcano commented:
There are some errors related to autocast

scruel committed:
Fix cast error.

scruel committed:
Fix autocast error.

The previous implementation called the `copy_` method, which implicitly cast the output dtype, consider the new impl won't need to call `copy_` anymore, so in this commit, we explicit cast the output tensor to act as the same as the old one.

scruel commented:
I got another failure about the stride mismatch problem, and we have already tried `.contiguous`, but the test wants to get the non-contiguous output for non-contiguous input.
After checking the code, we found that the method `compute_T18` will generate a contiguous output, so I think that we have to copy the result into a tensor created by `empty_like`, doing this also will fix the autocast error, and we won't need to call `to(dtype)` to fix it.


https://github.com/pytorch/pytorch/actions/runs/6531943267/job/17734968833?pr=110848

```
To execute this test, run the following from the base repo dir:
    PYTORCH_TEST_WITH_ROCM=1 python test_meta.py -k test_dispatch_symbolic_meta_outplace_all_strides_matrix_exp_cuda_float32

This message can be suppressed by setting PYTORCH_PRINT_REPRO_ON_FAILURE=0
- generated xml file: /var/lib/jenkins/pytorch/test/test-reports/python-pytest/test_meta/test_meta-b191d8473491dab4.xml -
=========================== short test summary info ============================
FAILED [0.0215s] test_meta.py::TestMetaCUDA::test_dispatch_symbolic_meta_outplace_all_strides_matrix_exp_cuda_float32 - RuntimeError: output 0: meta disagrees with real impl:
aten.linalg_matrix_exp.default(
  tensor(..., device='meta', size=(5, 5, 5)) stride=(1, 5, 25),
  
) = (
  tensor(..., device='meta', size=(5, 5, 5)) stride=(1, 5, 25)
)
but real stride was (25, 5, 1)

To execute this test, run the following from the base repo dir:
    PYTORCH_TEST_WITH_ROCM=1 python test_meta.py -k test_dispatch_symbolic_meta_outplace_all_strides_matrix_exp_cuda_float32

```

lezcano commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -2646,17 +2646,16 @@ Tensor mexp_impl(
         large_norm_subset,
         thetas[total_n_degs - 1]
-      );
+      ).to(a.dtype());
```
Is this really necessary?

scruel replied:
"Index put requires the source and destination dtypes match", this should be necessary.

lezcano commented:
Yeah, for those other cases, I think it's alright to modify the metafunction really.

scruel commented:
> Yeah, for those other cases, I think it's alright to modify the metafunction really.

For what other cases?

scruel commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -2646,17 +2646,16 @@ Tensor mexp_impl(
     a, norm,
     thetas[total_n_degs - 1]
-  );
+  ).to(a.dtype());
```
I believe we could, but currently I haven't checked it yet.

lezcano replied:
It'd be good if we could do that, yep.

scruel replied:
Oh, I misunderstood, `compute_T18` will only convert non-contiguous input to contiguous output, instead of changing the dtype.
The `matmul` operation (at line 2576) will change dtype within `torch.x.amp.autocast()` context.

JubilantJerry replied:
Looking at the test, it seems that due to AMP autocasting, the matrix multiplication of fp32 matrices can produce an fp16 result.
This happens in the main branch already when AMP is enabled, but at the end of the matrix_exp computation the fp16 matrix is copied into an fp32 result tensor, so the output appears to be fp32, and the test passes.
When AMP is not enabled, this added `.to(a.dtype())` is a no-op. `compute_T18` already preserves `dtype`.

scruel replied:
How about (Already tested, it has no significant performance impact):
```python
autocast_state = torch.is_autocast_enabled()
autocast_cpu_state = torch.is_autocast_cpu_enabled()

torch.set_autocast_enabled(False)
torch.set_autocast_cpu_enabled(False)

A = A @ A

torch.set_autocast_enabled(autocast_state)
torch.set_autocast_cpu_enabled(autocast_cpu_state)
```

lezcano commented:
The current failing tests.

scruel committed:
Fix try: keep incontiguous input

scruel commented:
> The current failing tests.

Oh, I see, however I don't I can fix that error, and I found that it already fixed in 18f0d3af7210f41dfb1ed17c65f103222bdb2563, so I just rebased the current branch, those tests should be able to pass now, let's just wait to see the result.

scruel committed:
Revert code clean: still need to call `to` before calling `index_put`

scruel commented:
I found that those irrelevant `TypeError` have been fixed in commit https://github.com/pytorch/pytorch/commit/18f0d3af7210f41dfb1ed17c65f103222bdb2563, so I rebased the current branch.

scruel committed:
Revert Contiguous option, skip `aten.linalg_matrix_exp.default` meta test.

JubilantJerry commented:
@scruel had been discussing ways of addressing two different test failures.
* Test failure 1 is a stride mismatch with the metafunction.
  * When the input is not contiguous, the metafunction's output is not contiguous, but the `matrix_exp` in this PR has a contiguous output.
* Test failure 2 is a type inconsistency when using AMP.
  * The `matrix_exp` in this PR sometimes autocasts fp32 inputs to fp16, and sometimes doesn't.

For test failure 1, the options are either to do an extra copy to change the memory format, or to change the [metafunction](https://github.com/pytorch/pytorch/blob/main/torch/_meta_registrations.py#L171-L174) to:
```
def linalg_matrix_exp(self):
    [...]
    return torch.empty_like(self, memory_format=torch.contiguous_format)
```

For test failure 2, the options are either to disable AMP inside `matrix_exp`, to do an extra copy to change the dtype, or to make both branches of `matrix_exp` autocast to fp16.

My opinion is we should change the metafunction to be contiguous, and to make both branches of `matrix_exp` autocast to fp16 when AMP wants it. I don't think we should disable the metafunction test case.

@lezcano What do you think about this way forward?

scruel commented:
@JubilantJerry For 1, can you be more clear about which metafunction you are referring to?
For 2, check https://github.com/pytorch/pytorch/pull/110848#discussion_r1361737964

JubilantJerry commented:
It's the one in the [metafunction](https://github.com/pytorch/pytorch/blob/main/torch/_meta_registrations.py#L171-L174) link.
In #110848 you suggest disabling AMP inside `matrix_exp`, which should work, but looking around I don't think any other function in PyTorch uses this approach. I might be wrong though.

scruel commented:
> It's the one in the [metafunction](https://github.com/pytorch/pytorch/blob/main/torch/_meta_registrations.py#L171-L174) link. In #110848 you suggest disabling AMP inside `matrix_exp`, which should work, but looking around I don't think any other function in PyTorch uses this approach. I might be wrong though.

Oh, now I got you, for 1, thanks for the code, I will change that and do the test, rather than disable it.
For 2, yep, it is kinda like a workaround, and not sure about if it will be right, I just got lucky to find it.
BTW, thank you for the clarification!

scruel commented:
> It's the one in the [metafunction](https://github.com/pytorch/pytorch/blob/main/torch/_meta_registrations.py#L171-L174) link. In #110848 you suggest disabling AMP inside `matrix_exp`, which should work, but looking around I don't think any other function in PyTorch uses this approach. I might be wrong though.

Oh, now I got you, for 1, thanks for the code, I will change that and do the test, rather than disable it.
For 2, yep, it is kinda like a workaround, and not sure about if it will be right, I just got lucky to find it. And if you think it's ok to output different dtype within AMP autocast context, then we still need to change the test code.
BTW, thank you for your clarification, I am also confused about these two errors until I know what I am doing. :)

scruel commented:
```python
def linalg_matrix_exp(self):
    [...]
    return torch.empty_like(self, memory_format=torch.contiguous_format)
```

By modifying this, the test case still will be failed:
```
E
======================================================================
ERROR: test_dispatch_symbolic_meta_outplace_all_strides_matrix_exp_cuda_float32 (__main__.TestMetaCUDA)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/scruel/scp/pytorch/torch/testing/_internal/common_utils.py", line 2450, in wrapper
    method(*args, **kwargs)
  File "/home/scruel/scp/pytorch/torch/testing/_internal/common_utils.py", line 2450, in wrapper
    method(*args, **kwargs)
  File "/home/scruel/scp/pytorch/torch/testing/_internal/common_device_type.py", line 428, in instantiated_test
    raise rte
  File "/home/scruel/scp/pytorch/torch/testing/_internal/common_device_type.py", line 415, in instantiated_test
    result = test(self, **param_kwargs)
  File "/home/scruel/scp/pytorch/torch/testing/_internal/common_device_type.py", line 908, in test_wrapper
    return test(*args, **kwargs)
  File "/home/scruel/scp/pytorch/torch/testing/_internal/common_utils.py", line 1164, in wrapper
    fn(*args, **kwargs)
  File "/home/scruel/scp/pytorch/torch/testing/_internal/common_utils.py", line 1621, in wrapper
    fn(*args, **kwargs)
  File "/home/scruel/scp/pytorch/torch/testing/_internal/common_device_type.py", line 1084, in only_fn
    return fn(slf, *args, **kwargs)
  File "/home/scruel/scp/pytorch/test/test_meta.py", line 1239, in test_dispatch_symbolic_meta_outplace_all_strides
    self._run_dispatch_meta_test(device, dtype, op, symbolic_meta=True, inplace=False, all_stride_variants=True)
  File "/home/scruel/scp/pytorch/test/test_meta.py", line 1196, in _run_dispatch_meta_test
    expected = func(*args, **kwargs)
  File "/home/scruel/scp/pytorch/test/test_meta.py", line 1040, in __torch_dispatch__
    expected = run_meta_crossref(
  File "/home/scruel/scp/pytorch/test/test_meta.py", line 584, in run_meta_crossref
    assert_ref_meta_equal(test_case, func, meta_rs, rs, lambda msg: f"""\
  File "/home/scruel/scp/pytorch/test/test_meta.py", line 395, in assert_ref_meta_equal
    test_assert(same_strides, f"but real stride was {r.stride()}")
  File "/home/scruel/scp/pytorch/test/test_meta.py", line 383, in test_assert
    raise RuntimeError(f"output {i}: {msg_callable(msg)}")
RuntimeError: output 0: meta disagrees with real impl:
aten.linalg_matrix_exp.default(
  tensor(..., device='meta', size=(5, 5)) stride=(1, 5),
  
) = (
  tensor(..., device='meta', size=(5, 5)) stride=(5, 1)
)
but real stride was (1, 5)


To execute this test, run the following from the base repo dir:
     python test/test_meta.py -k test_dispatch_symbolic_meta_outplace_all_strides_matrix_exp_cuda_float32
```

scruel committed:
Fix failing test case.

lezcano commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -2510,32 +2517,74 @@ Tensor compute_T18(const Tensor& A) {
+  for (int64_t i = 0; i < section_numel; ++i) {
+    for (int64_t j = 0; j < pts[i]; j++) {
+      // To avoid AMP autocasting caused by at::matmul
+      auto acc_out = at::empty_like(acc);
```
Even better, you can declare it outside of this loop to reuse it for all the matmul_out calls

JubilantJerry replied:
Since the output of the matmul is used as the input in the next loop iteration, I think there has to be at least 2 temporary matrices outside the loop to do that, otherwise the matmul would have to write out to its own input, which I think is not allowed?

scruel replied:
If we can do that, then we can simply remove `acc_out`.

scruel replied:
The result would be wrong if we move the declaration outside the loop.

lezcano replied:
Right. It was just a bit confusing the aliasing between acc and acc_out, but yep

scruel replied:
BTW, instead of using at::matmul_out, I think we can use _matmul_impl here.

lezcano reviewed (commented):
Yep, both changes LGTM, just a minor to avoid creating a few unnecessary matrices.

lezcano commented:
@pytorchbot merge

lezcano commented:
@pytorchbot merge -f "Unrelated error"
